File: js/controllers/items.js

var model = undefined;

exports.context = function(server, path) {
    if (!server)
        done('has to provide a restify server object');
        
    var context = "/items";
    if (path)
        context = path + context;
        
    server.get(context + '/', this.list);
    server.get(context + '/:id', this.read);
    server.post(context + '/', this.save);
    server.del(context + '/:id', this.destroy);
};

exports.model = function(param) {
    model = param;
}


exports.list = function(req, res, next) {
    model.listAll(function(err, items) {
        if (err) {
            res.status(500);
            res.json({
                error: err
            })
        }
        else {
            if (items) {
                //XXX no pagination yet
                var page = { 
                  "currentPage" : 1,
                  "list" : items,
                  "pageSize" : 10,
                  "sortDirections" : "asc",
                  "sortFields" : "id",
                  "totalResults" : items.length
                };
                res.json(page);
            }
            else {
                // XXX how the jee_api returns errors?
                res.json({
                    error: "Can't retrieve items"
                });
            }
        }
    })
};

exports.read = function(req, res, next) {
    var key = req.params.id;
    model.read(key, function(err, item) {
        if (err) {
            res.status(500);
            res.json({
                error: err
            })
        }
        else {
            if (item) {
                res.json(item);
            }
            else {
                // XXX how the jee_api returns errors?
                res.json({
                    error: "Can't retrieve items"
                });
            }
        }
    })
};


exports.save = function(req, res, next) {
    if (req.params.id) {
        model.update(req.params.id, req.params.description, req.params.done, function(err, item) {
            if (err) {
                res.status(500);
                res.json({
                    error: err
                })
            }
            else {
                res.json(item);
            }
        });
    }
    else {
        model.create(req.params.description, req.params.done, function(err, item) {
            if (err) {
                res.status(500);
                res.json({
                    error: err
                })
            }
            else {
                res.json(item);
            }
        });
    }
};


exports.destroy = function(req, res, next) {
    if (req.params.id) {
        model.destroy(req.params.id, function(err, item) {
            if (err) {
                res.status(500);
                res.json({
                    error: err
                })
            }
            else {
                //XXX jee_api does NOT return item on delete
                res.json(item);
            }
        });
    }
}

File: apps/ruby/routes/items.rb
# encoding: UTF-8

get '/todo/api/items' do

    total = Item.count()
    
    sortFields = params['sortFields']
    sortDirections = params['sortDirections']
    page = Integer(params['page'])
    
    items = Item.all

    if (sortFields == 'id')
        items = items.order(id: (sortDirections == 'desc') ? :desc : :asc)
    elsif (sortFields == 'description')
        items = items.order(description: (sortDirections == 'desc') ? :desc : :asc)
    elsif (sortFields == 'done')
        items = items.order(done: (sortDirections == 'desc') ? :desc : :asc)
    elsif
        items = items.order(id: :asc)
    end        

    page = (page > 0) ? page : 1 
    items = items.offset(10 * (page - 1)).limit(10)

    response = { 
        "currentPage" => page,
        "list" => items,
        "pageSize" => 10,
        "sortDirections" => sortDirections,
        "sortFields" => sortFields,
        "totalResults" => total
    }
    return response.to_json
end

get '/todo/api/items/:id' do
    item ||= Item.find(params[:id]) || halt(404)
    return item.to_json
end

post '/todo/api/items' do
    body = JSON.parse request.body.read
    if (body['id'])
    item ||= Item.find(body['id']) || halt(404)
        item.description = body['description']
        item.done = body['done']
        item.save
    else 
        item ||= Item.create(description: body['description'], done: body['done'],) || halt(404)
    end
    return item.to_json
end

delete '/todo/api/items/:id' do
    item ||= Item.find(params[:id]) || halt(404)
    item.destroy
    return item.to_json
end


